src/builder.rs

```diff
@@ -0,0 +1,472 @@
+                rk,
+                cm_new.into(),
+                encrypted_note,
+                cv_net,
+                (
+                    self.spend.dummy_sk.as_ref().map(SpendAuthorizingKey::from),
+                    ak,
+                ),
+            ),
+            Circuit {},
+        )
+    }
+}
+
+/// A builder that constructs a [`Bundle`] from a set of notes to be spent, and recipients
+/// to receive funds.
+pub struct Builder {
+    spends: Vec<SpendInfo>,
+    recipients: Vec<RecipientInfo>,
+    flags: Flags,
+    anchor: Anchor,
+}
+
+impl Builder {
+    pub fn new(flags: Flags, anchor: Anchor) -> Self {
+        Builder {
+            spends: vec![],
+            recipients: vec![],
+            flags,
+            anchor,
+        }
+    }
+
+    /// Adds a note to be spent in this transaction.
+    ///
+    /// Returns an error if the given Merkle path does not have the required anchor for
+    /// the given note.
+    pub fn add_spend(
+        &mut self,
+        fvk: FullViewingKey,
+        note: Note,
+        merkle_path: MerklePath,
+    ) -> Result<(), &'static str> {
+        if !self.flags.spends_enabled() {
+            return Err("Spends are not enabled for this builder");
+        }
+
+        // Consistency check: all anchors must be equal.
+        let cm = note.commitment();
+        // TODO: Once we have tree logic.
+        // let path_root: bls12_381::Scalar = merkle_path.root(cmu).into();
+        // if path_root != anchor {
+        //     return Err(Error::AnchorMismatch);
+        // }
+
+        self.spends.push(SpendInfo {
+            dummy_sk: None,
+            fvk,
+            note,
+            merkle_path,
+        });
+
+        Ok(())
+    }
+
+    /// Adds an address which will receive funds in this transaction.
+    pub fn add_recipient(
+        &mut self,
+        ovk: Option<OutgoingViewingKey>,
+        recipient: Address,
+        value: NoteValue,
+        memo: Option<()>,
+    ) -> Result<(), &'static str> {
+        if !self.flags.outputs_enabled() {
+            return Err("Outputs are not enabled for this builder");
+        }
+
+        self.recipients.push(RecipientInfo {
+            ovk,
+            recipient,
+            value,
+            memo,
+        });
+
+        Ok(())
+    }
+
+    /// Builds a bundle containing the given spent notes and recipients.
+    ///
+    /// This API assumes that none of the notes being spent are controlled by (threshold)
+    /// multisignatures, and immediately constructs the bundle proof.
+    fn build(
+        mut self,
+        mut rng: impl RngCore,
+        pk: &ProvingKey,
+    ) -> Result<Bundle<Unauthorized>, Error> {
+        // Pair up the spends and recipients, extending with dummy values as necessary.
+        //
+        // TODO: Do we want to shuffle the order like we do for Sapling? And if we do, do
+        // we need the extra logic for mapping the user-provided input order to the
+        // shuffled order?
```

Comment:
Could the randomization be left to be a wallet-level concern? What would the randomization buy us here?

Reply:
It can't quite be a wallet-level concern because the bundle builder encapsulates the logic for dummy notes, amongst which any real notes would be randomized.

What randomization removes is any information about what the position of a spend or output within an Action might imply. For example, if outputs are always in order, then if there are three actions in a bundle and you receive a new note as the output of the second action, then you can reasonably infer that the third action's output is most likely change to the sender, _and_ there was another recipient (in the first action's output). If outputs are randomized, then that is not distinguishable from the sender's wallet just performing note management and sending itself two notes.